libcontainer/dmz/overlayfs_linux.go

```diff
@@ -0,0 +1,108 @@
+	}
+	return os.NewFile(uintptr(fd), "fscontext:"+fsName), nil
+}
+
+func fsmount(ctx *os.File, flags, mountAttrs int) (*os.File, error) {
+	// Make sure we always set O_CLOEXEC.
+	flags |= unix.FSMOUNT_CLOEXEC
+	fd, err := unix.Fsmount(int(ctx.Fd()), flags, mountAttrs)
+	if err != nil {
+		return nil, os.NewSyscallError("fsmount "+ctx.Name(), err)
+	}
+	runtime.KeepAlive(ctx) // make sure fd is kept alive while it's used
+	return os.NewFile(uintptr(fd), "fsmount:"+ctx.Name()), nil
+}
+
+func escapeOverlayLowerDir(path string) string {
+	// If the lowerdir path contains ":" we need to escape them, and if there
+	// were any escape characters already (\) we need to escape those first.
+	return strings.ReplaceAll(strings.ReplaceAll(path, `\`, `\\`), `:`, `\:`)
+}
+
+// sealedOverlayfs will create an internal overlayfs mount using fsopen() that
+// uses the directory containing the binary as a lowerdir and a temporary tmpfs
+// as an upperdir. There is no way to "unwrap" this (unlike MS_BIND+MS_RDONLY)
+// and so we can create a safe zero-copy sealed version of /proc/self/exe.
+func sealedOverlayfs(binPath, tmpDir string) (_ *os.File, Err error) {
+	// Try to do the superblock creation first to bail out early if we can't
+	// use this method.
+	overlayCtx, err := fsopen("overlay", unix.FSOPEN_CLOEXEC)
+	if err != nil {
+		return nil, err
+	}
+	defer overlayCtx.Close()
+
+	// binPath is going to be /proc/self/exe, so do a readlink to get the real
+	// path. overlayfs needs the real underlying directory for this protection
+	// mode to work properly.
+	if realPath, err := os.Readlink(binPath); err == nil {
+		binPath = realPath
+	}
+	binLowerDirPath, binName := filepath.Split(binPath)
+	// Escape any ":"s or "\"s in the path.
+	binLowerDirPath = escapeOverlayLowerDir(binLowerDirPath)
+
+	// Overlayfs requires two lowerdirs in order to run in "lower-only" mode,
+	// where writes are completely blocked. Ideally we would create a dummy
+	// tmpfs for this, but it turns out that overlayfs doesn't allow for
+	// anonymous mountns paths.
+	// NOTE: I'm working on a patch to fix this but it won't be backported.
+	dummyLowerDirPath := escapeOverlayLowerDir(tmpDir)
+
+	// Configure the lowerdirs. The binary lowerdir needs to be on the top to
+	// ensure that a file called "runc" (binName) in the dummy lowerdir doesn't
+	// mask the binary.
+	lowerDirStr := binLowerDirPath + ":" + dummyLowerDirPath
+	if err := unix.FsconfigSetString(int(overlayCtx.Fd()), "lowerdir", lowerDirStr); err != nil {
+		return nil, fmt.Errorf("fsconfig set overlayfs lowerdir=%s: %w", lowerDirStr, err)
+	}
+
+	// Get an actual handle to the overlayfs.
+	if err := unix.FsconfigCreate(int(overlayCtx.Fd())); err != nil {
+		return nil, os.NewSyscallError("fsconfig create overlayfs", err)
+	}
+	overlayFd, err := fsmount(overlayCtx, unix.FSMOUNT_CLOEXEC, unix.MS_RDONLY|unix.MS_NODEV|unix.MS_NOSUID)
+	if err != nil {
+		return nil, err
+	}
+	defer overlayFd.Close()
```

Comment:
I guess the mount object will continue to exist after this close, but it will be freed by the kernel when we exec into the container process, and then /proc/self/exe is not runc anymore and then it is freed? Is that the life-time of the mount object?

Reply:
Thanks!

Reply:
One more question: can this be abused, maybe? They can't overwrite the real runc binary, but they can keep the fd open and therefore keep the mount alive. Is there something an attacker can do with that? I don't know how much mem this takes, or if piling up (like launching a lot of containers that keep the detached mount alive) can create some undesired effect?

Because any unprivileged container will be able to do it.

Reply:
Keeping the mount alive would waste some small amount of kernel memory (but you would also need to be able to spawn many containers to do this for appreciable amounts of memory) and because it keeps the anonymous namespace alive you could end up hitting the per-user limit of mount namespaces in `/proc/sys/user/max_mnt_namespaces` (if you do this for many containers) which would stop you from being able to start more containers. However, an attacker being able to arbitrarily start hundreds of thousands of containers (on my system the default limit is ~200k) already would cause you to exhaust this limit.

The only approach I can think of that wouldn't cause any extra resource usage if an attacker pins the file descriptor is `memfd-bind` but because we can't bind-mount the underlying `memfd` (and instead have to bind-mount the magic-link pointing to it) that is not workable for most people.

There are some fairly high time-complexity algorithms in the mount infrastructure of Linux (`umount` and mount propagation in particular) but those are all related to large number of mounts in a namespace or lots of namespaces with shared propagation. Anonymous mount namespaces aren't relevant for those because they are constructed as a single namespace with a single mount that is completely isolated from every other namespace and mount tree.

Reply:
Perfect, let's give this a try!